keps/sig-auth/3257-trust-anchor-sets/README.md

```diff
@@ -0,0 +1,1328 @@
+they are updated, workloads will need to receive the updates fairly quickly
+(within 5 minutes across the whole cluster), to accommodate emergency rotation
+of trust anchors for a private CA.
+
+Security: Should individual trust anchor set entries designate an OSCP endpoint
+to check for certificate revociation?  Or should we require the URL to be
+embedded in the issued certificates?
+
+## Design Details
+
+### ClusterTrustBundle Object
+
+The ClusterTrustBundle object allows signers to publish their trust anchors for
+consumption by workloads, whether directly, via Kubelet, or otherwise.
+
+ClusterTrustBundle objects may optionally be associated with a signer using the
+`.spec.signerName` field.  Setting the `.spec.signerName` field enables
+additional admission control and access control for the ClusterTrustBundle
+object, as detailed in the [Access Control](#access-control) section.
+
+While multiple ClusterTrustBundle objects may be associated with a signer, each
+is a completely independent expression of the roots of trust for the signer.  No
+effort is made to unify trust anchors from different ClusterTrustBundle objects.
+The primary use case for multiple ClusterTrustBundle objects for a single signer
+is [canarying](#canarying-changes-to-a-clustertrustbundle).
+
+All of the new objects and fields described in this section are gated by the
+ClusterTrustBundle kube-apiserver feature gate.  Unless the feature gate is
+enabled, usage of these alpha objects and fields will be rejected by
+kube-apiserver.
+
+#### API Object Definition
+
+```go
+// ClusterTrustBundle packages up trust anchors.
+//
+// Each ClusterTrustBundle object is optionally owned by a particular signer, which
+// updates the trust anchors as the roots of trust for that signer change.
+//
+// A ClusterTrustBundle object wraps up multiple trust anchor sets, each of
+// which contains one or more trust anchors (individual
+// certificates).  Most use cases will only require a single trust
+// anchor set, conventionally named `default`.  Multiple uniquely-named trust
+// anchor sets will likely only be required in mutual TLS use cases with
+// federation between multiple meshes.
+type ClusterTrustBundle struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // The spec of the object.
+    Spec ClusterTrustBundleSpec `json:"spec"`
+
+    // The status of the object.
+    Status ClusterTrustBundleStatus `json:"status"`
+}
+
+// ClusterTrustBundleSpec is the spec subfield of a ClusterTrustBundle object.
+type ClusterTrustBundleSpec struct {
+  // The name of the associated signer.
+  //
+  // +optional
+  SignerName string `json:"signerName,omitempty"`
+}
+
+// ClusterTrustBundleStatus is the status subfield of a ClusterTrustBundle object.
+type ClusterTrustBundleStatus struct {
+    // The list of trust anchor sets.
+    TrustAnchorSets []TrustAnchorSet `json:"trustAnchorSets"`
```

Comment:
Because of the mTLS federation use case --- we may have an unbounded number of sets if we operate company A that's federating with companies B, C, ... Z.  My server doesn't know which trust anchor set to use until it inspects the certificate presented by the client.  The choice of which trust anchor set to use might involve looking at URI SANs, CNs, or even OUs (which can contain nastiness like apostrophes, slashes, etc).

I don't want to force signers to come up with a naming convention for encoding these arbitrary properties into ClusterTrustBundle names, which will be prone to spoofing.  Instead, I'd like to give the signer a 100% safe, no-encoding-required string to place whatever data is needed for the trust lookup.

Reply:
enj@ also requested some fleshing out of the mTLS use case.  I've added a placeholder section for a user story.

Reply:
I decided to boot the mTLS use case.  Everything I think is needed can be handled for now using annotations on the ClusterTrustBundle object.